### PR TITLE
WIP: initial jobs/threads heuristic

### PR DESCRIPTION
**Pull request by kergoth**
JIRA: MELMR-1124

### CODE REVIEW

File: meta-mel/conf/distro/include/jobs-heuristic.inc [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                    d.setVar('JOBS_HEURISTIC_ERROR',
+                             'Failed to evaluate jobs (%s) in JOBS_HEURISTIC_THRESHOLDS entry %s: %s' % (
+                                jobs, entry, exc))
+                    return
+                break
+        else:
+            return
+
+    bb.debug(1, 'Memory per CPU core is %s, adjusting BB_NUMBER_THREADS=%s, PARALLEL_MAKE=%s' % (mem_per_cpu, threads, jobs))
+    if not d.getVar('BB_NUMBER_THREADS'):
```
Comment (abelal) [inline meta-mel/conf/distro/include/jobs-heuristic.inc:68]: isn't this set from the bitbake.conf already in the default case?

File: meta-mel/conf/distro/include/jobs-heuristic.inc [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        d.setVar('BB_NUMBER_THREADS', str(max(threads, 1)))
+    if not d.getVar('PARALLEL_MAKE'):
```
Comment (abelal) [inline meta-mel/conf/distro/include/jobs-heuristic.inc:70]: and so is this?

File: meta-mel/conf/distro/include/jobs-heuristic.inc (hunk 1 above)
Comment (kergoth) [reply meta-mel/conf/distro/include/jobs-heuristic.inc:68]: Yes, good point, I'll update it.

File: meta-mel/conf/distro/include/jobs-heuristic.inc (hunk 2 above)
Comment (kergoth) [reply meta-mel/conf/distro/include/jobs-heuristic.inc:70]: Same.

File: meta-mel/conf/distro/include/jobs-heuristic.inc [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                                jobs, entry, exc))
+                    return
+                break
+        else:
+            return
+
+    bb.debug(1, 'Memory per CPU core is %s, adjusting BB_NUMBER_THREADS to %s and PARALLEL_MAKE to %s' %
+                  (mem_per_cpu, threads, jobs))
+    d.setVar('BB_NUMBER_THREADS', str(max(threads, 1)))
+    d.setVar('PARALLEL_MAKE', '-j %s' % max(jobs, 1))
```
Comment (abelal) [inline meta-mel/conf/distro/include/jobs-heuristic.inc:70]: Wouldn't this override whatever we set in local.conf?

File: conf/local.conf.sample [hunk 4]
```diff
@@ -91,10 +91,10 @@ USER_FEATURES += "~3g"
 USER_FEATURES += "~nfc"
 
 # How many tasks bitbake should run in parallel
-BB_NUMBER_THREADS ?= "${@int(oe.utils.cpu_count() * 1.5)}"
+# BB_NUMBER_THREADS ?= "${@int(oe.utils.cpu_count())}"
 
 # How many jobs 'make' should run in parallel when compiling
-PARALLEL_MAKE ?= "-j ${@oe.utils.cpu_count() * 2}"
+# PARALLEL_MAKE ?= "-j ${@oe.utils.cpu_count()}"
```
Comment (abelal) [inline conf/local.conf.sample:97]: Drop the space after the # in these variable commenting cases? I guess that's how we do it for such stuff.

File: meta-mel/conf/distro/include/jobs-heuristic.inc (hunk 3 above)
Comment (kergoth) [reply meta-mel/conf/distro/include/jobs-heuristic.inc]: Updated to set intermediate variables which are then referenced by mel.conf with ?=. This ensures that local.conf would be preferred, as it's parsed prior to mel.conf, which is parsed prior to the defaults in bitbake.conf.

File: conf/local.conf.sample (hunk 4 above)
Comment (kergoth) [reply conf/local.conf.sample]: Fixed, thanks

File: meta-mel/conf/distro/include/jobs-heuristic.inc [hunk 5]
```diff
@@ -0,0 +1,95 @@
+# Set bitbake threads and make jobs based upon available memory per CPU core.
+# Based on https://elinux.org/images/d/d4/Goulart.pdf, a talk done by Mario Domenech
```
Comment (abelal) [inline meta-mel/conf/distro/include/jobs-heuristic.inc:2]: So we've implemented the algo as per page 50 of the mentioned study? If yes, that looks a bit different to me. Comments to follow on the exact scenarios.

File: meta-mel/conf/distro/include/jobs-heuristic.inc [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+# Goulart and Mikko Rapeli from BMW CAR IT GMBH.
+#
+# MEM_PER_CPU_MINIMUM,THREADS,JOBS,ACTION
+JOBS_HEURISTIC_THRESHOLDS ?= "\
+    7.9,cpus,cpus*1.5, \
```
Comment (abelal) [inline meta-mel/conf/distro/include/jobs-heuristic.inc:7]: for 8G or greater per cpu the study uses threads=cpus pmake=cpus, we've pmake=cpus*1.5 here.

File: meta-mel/conf/distro/include/jobs-heuristic.inc [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    3.9,cpus,cpus, \
```
Comment (abelal) [inline meta-mel/conf/distro/include/jobs-heuristic.inc:8]: for greater than 4G/cpu the study uses threads=cpus pmake=cpus/2, we've pmake=cpus

File: meta-mel/conf/distro/include/jobs-heuristic.inc [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    1.9,cpus//2,cpus//2,WARN \
```
Comment (abelal) [inline meta-mel/conf/distro/include/jobs-heuristic.inc:9]: aligned with the study here

File: meta-mel/conf/distro/mel.conf [hunk 9]
```diff
@@ -47,6 +47,15 @@ WARN_TO_ERROR_QA ??= ""
 WARN_QA:remove = "${WARN_TO_ERROR_QA}"
 ERROR_QA:append = " ${WARN_TO_ERROR_QA}"
 
+# Adapt default threads and jobs based on available memory
+require conf/distro/include/jobs-heuristic.inc
+BB_NUMBER_THREADS ?= "${BB_NUMBER_THREADS_HEURISTIC}"
+PARALLEL_MAKE ?= "${PARALLEL_MAKME_HEURISTIC}"
```
Comment (abelal) [inline meta-mel/conf/distro/mel.conf:53]: typo here
PARALLEL_MAK**M**E_HEURISTIC

**Comment by kergoth** [comment]
@abelal I think I tweaked it based on our current default values which are more aggressive than the defaults in bitbake.conf, we'll have to decide if we want to continue our more aggressive defaults, which tend to hit problems in constrained environments, or switch to upstream more conservative defaults and align with the study. I'm inclined to the latter, in which case it should be changed based on your comments. Thoughts, @Noor-Ahsan ?

**Comment by abelal** [comment]
> @abelal I think I tweaked it based on our current default values which are more aggressive than the defaults in bitbake.conf, we'll have to decide if we want to continue our more aggressive defaults, which tend to hit problems in constrained environments, or switch to upstream more conservative defaults and align with the study. I'm inclined to the latter, in which case it should be changed based on your comments. Thoughts, @Noor-Ahsan ?

I'd agree to this.

**Comment by Noor-Ahsan** [comment]
> > @abelal I think I tweaked it based on our current default values which are more aggressive than the defaults in bitbake.conf, we'll have to decide if we want to continue our more aggressive defaults, which tend to hit problems in constrained environments, or switch to upstream more conservative defaults and align with the study. I'm inclined to the latter, in which case it should be changed based on your comments. Thoughts, @Noor-Ahsan ?
> 
> I'd agree to this.

I would agree to this. our defautls is based on study that we did at early stage of the product. After that we did not update it.

**Comment by kergoth** [comment]
Migrated to https://github.com/MentorEmbedded/meta-sokol-flex/pull/13

**Pull request closed without merging**